p2p/protocol/head/head.go

```diff
@@ -63,7 +63,22 @@ func QueryRootCid(ctx context.Context, host host.Host, topic string, peerID peer
 				if err != nil {
 					return nil, err
 				}
-				return gostream.Dial(ctx, host, peerID, deriveProtocolID(topic))
+				conn, err := gostream.Dial(ctx, host, peerID, deriveProtocolID(topic))
+				if err != nil {
+					// If protocol ID is wrong, then try the old "double-slashed" protocol ID.
+					//
+					// TODO: remove this code when all providers have upgraded.
+					if err.Error() != "protocol not supported" {
```

Comment:
I suspect the error returned is [this](https://github.com/multiformats/go-multistream/blob/master/client.go#L18), which would make it possible to check the type instead of message.

Reply:
Thank you. I thought it might be [this one](https://github.com/libp2p/go-libp2p/blob/master/p2p/protocol/identify/peer_loop.go#L21) which is not exported. After testing, looks like you are correct.